pkg/statistics/handle/globalstats/topn.go

```diff
@@ -138,7 +262,173 @@ func MergeGlobalStatsTopNByConcurrency(
+		for histPos, found := cur.affectedTopNs.NextClear(0); found; histPos, found = cur.affectedTopNs.NextClear(histPos + 1) {
+			affectedHist = append(affectedHist, int(histPos))
+		}
+		// Hacking skip.
+		if uint32(len(finalTopNs)) >= n {
+			maxPossible := int64(0)
+			for _, histPos := range affectedHist {
+				maxPossible += maxPossibleAdded[histPos]
+			}
+			// The maximum possible added value still cannot make it replace the smallest topn.
+			if maxPossible+int64(cur.item.Count) < int64(finalTopNs[0].Count) {
+				skipCount++
+				remainedTopNs = append(remainedTopNs, statistics.TopNMeta{Encoded: cur.item.Encoded, Count: cur.item.Count})
+				// Set the cur maintained to the next value.
+				cur.item.Encoded = nextVal.Encoded
+				cur.item.Count = nextVal.Count
+				cur.affectedTopNs.ClearAll()
+				cur.affectedTopNs.Set(position)
+				return nil
+			}
+		}
+		for _, histPos := range affectedHist {
+			histRemoveCnt++
+			// Remove the value from the hist and add it into the current maintained value.
+			cur.item.Count += uint64(histIters[histPos].remove(&d))
+		}
+		// Size reaches the n, maintaining the heap.
+		if finalTopNs.Len() == int(n) {
+			if finalTopNs[0].Count < cur.item.Count {
+				remainedTopNs = append(remainedTopNs, finalTopNs[0])
+				finalTopNs[0].Encoded = cur.item.Encoded
+				finalTopNs[0].Count = cur.item.Count
+				heap.Fix(&finalTopNs, 0)
+			} else {
+				remainedTopNs = append(remainedTopNs, cur.item)
+			}
+		} else {
+			// Otherwise the heap is not fulfilled.
+			finalTopNs = append(finalTopNs, statistics.TopNMeta{Encoded: cur.item.Encoded, Count: cur.item.Count})
+			if finalTopNs.Len() == int(n) {
+				heap.Init(&finalTopNs)
+			}
+		}
+		// Set the cur maintained to the next value.
+		cur.item.Encoded = nextVal.Encoded
+		cur.item.Count = nextVal.Count
+		cur.affectedTopNs.ClearAll()
+		cur.affectedTopNs.Set(position)
+		return nil
+	}
+	for {
+		if err := killer.HandleSignal(); err != nil {
+			return nil, nil, nil, err
+		}
+		if mergingHeap.Len() == 0 {
+			break
+		}
+		step++
+		head := mergingHeap[0]
+		headTopN := head.item
+		if head.nextPosInTopN < topNs[head.idx].Num() {
+			head.item = &topNs[head.idx].TopN[head.nextPosInTopN]
+			head.nextPosInTopN++
+			heap.Fix(&mergingHeap, 0)
+		} else {
+			heap.Pop(&mergingHeap)
+		}
+		// Init the cur when we first enter the heap.
+		if firstTime {
+			cur.item.Encoded = headTopN.Encoded
+			cur.item.Count = headTopN.Count
+			cur.affectedTopNs.Set(uint(head.idx))
+			firstTime = false
+			continue
+		}
+		cmp := bytes.Compare(cur.item.Encoded, headTopN.Encoded)
+		// The heap's head move forward.
+		if cmp < 0 {
+			err := checkTheCurAndMoveForward(headTopN, uint(head.idx))
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			continue
+		}
+		// The cmp result cannot be 1 because the value is strictly increasing.
+		// Here is cmp == 0.
+		cur.item.Count += headTopN.Count
+		cur.affectedTopNs.Set(uint(head.idx))
+	}
+	{
```

Comment:
Why do we need a separate block here?

Reply:
The block is added before i extract that func. It's a longer one compared at that time.